File: Shortest Path/Floyd-Warshall/2458번 - 키 순서/2458.cpp
//2458번: 키 순서
#include <iostream>

using namespace std;

int n, m;
int stu[501][501];

void input() {
    cin >> n >> m;
    for(int i = 0, a, b; i < m; i++) {
        cin >> a >> b;
        stu[a][b] = 1;
        stu[b][a] = -1;
    }
}

void solve() {
    for(int k = 1; k <= n; k++) {
        for(int i = 1; i <= n; i++) {
            for(int j = 1; j <= n; j++) {
                if(stu[i][k] == stu[k][j] && stu[i][k] != 0) stu[i][j] = stu[i][k];
            }
        }
    }
    
    int ans = 0;
    
    for(int i = 1; i <= n; i++) {
        int cnt = n - 1;
        for(int j = 1; j <= n; j++) {
            if(stu[i][j] != 0) cnt--;
        }
        if(cnt == 0) ans++;
    }
    cout << ans << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Shortest Path/Floyd-Warshall/2458번 - 키 순서/2458.py
import sys

input = sys.stdin.readline

n, m = map(int, input().split())
stu = [[0] * n for _ in range(n)]
for _ in range(m):
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    stu[a][b] = 1

def solve():
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if stu[i][k] and stu[k][j]:
                    stu[i][j] = 1
    
    ans = 0
    for i in range(n):
        if sum(stu[i]) + sum(st[i] for st in stu) == n - 1:
            ans += 1
    print(ans)

if __name__ == '__main__':
    solve()